src/esy.ml

```diff
@@ -91,10 +89,19 @@ module Discover = struct
          (Path.toString dir);
        [||]
      | Ok res -> res)
-    >>= fun files ->
-    files
-    |> Js.Array.map (parseFile dir)
-    |> Promise.all |> Promise.map foldResults
+    >>= fun files -> files |> Js.Array.map (parseFile dir) |> Promise.all
+
+  let getUpstreamDirs dir =
+    let dirList = dir |> Path.split in
+    dirList |> List.length
+    |. Belt.List.makeBy (fun i ->
+           dirList |. Belt.List.take (i + 1) |. Belt.Option.map Path.concat)
+    |. Belt.List.keepMap (fun x -> x)
```

Comment:
This is a bit convoluted. It would be simpler if you'd introduce a primitive like this:

```
module Path : sig
  type t
  val parent : t -> t option
end
```

And then recurse with it until you reach root.

Reply:
good idea, will do

Reply:
Just lurking over here. Feel free to use this little one-file library I made called `Fp` that helps with a lot of this stuff if you just need a self-contained util: https://github.com/facebookexperimental/reason-native/tree/master/src/fp

Currently requires that you normalize platform paths on the way in, but we'll be adding those utility features directly.

Reply:
Awesome! I guess we can mostly replace `Path` with `Fp` then. But I think I will do it in a separate PR as it will touch a lot of stuff

Reply:
Just as an aside, we don't permit any reason sources in this repository. So you will have to port this module to OCaml first.